File: python/examples/product.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from __future__ import print_function, division

from goofit import *
import numpy as np
import sys

print_goofit_info()

xdata = np.random.exponential(size=100000)
xvar = Observable("xvar", 0, np.max(xdata) + 1)

ydata = np.random.exponential(size=100000)
yvar = Observable("yvar", 0, np.max(ydata) + 1)

varList = (xvar,yvar)

data = UnbinnedDataSet(xvar,yvar)
data.from_matrix([xdata, ydata])

alpha_x = Variable("alpha_x", -2.4, 0.1, -10, 10)
alpha_y = Variable("alpha_y", -1.1, 0.1, -10, 10)

exp_x = ExpPdf("exp_x", xvar, alpha_x)
exp_y = ExpPdf("exp_x", yvar, alpha_y)
product = ProdPdf("product", (exp_x, exp_y))

product.setData(data)
fitter = FitManager(product)
fitter.fit()

sys.exit(int(int(fitter)))

File: examples/product/product.cpp
#include <goofit/Application.h>
#include <goofit/FitManager.h>
#include <goofit/PDFs/basic/ExpPdf.h>
#include <goofit/PDFs/combine/ProdPdf.h>
#include <goofit/UnbinnedDataSet.h>
#include <goofit/Variable.h>

using namespace std;
using namespace GooFit;

int main(int argc, char **argv) {
    GooFit::Application app("Product example", argc, argv);

    GOOFIT_PARSE(app);

    Observable xvar{"xvar", 0, log(1 + RAND_MAX / 2)};
    Observable yvar{"yvar", 0, log(1 + RAND_MAX / 2)};
    vector<Observable> varList = {xvar, yvar};
    UnbinnedDataSet data{varList};

    for(int i = 0; i < 100000; ++i) {
        xvar.setValue(xvar.getUpperLimit() - log(1 + rand() / 2));
        yvar.setValue(yvar.getUpperLimit() - log(1 + rand() / 2));
        data.addEvent();
    }

    Variable alpha_x{"alpha_x", -2.4, 0.1, -10, 10};
    Variable alpha_y{"alpha_y", -1.1, 0.1, -10, 10};

    ExpPdf exp_x{"exp_x", xvar, alpha_x};
    ExpPdf exp_y{"exp_y", yvar, alpha_y};
    ProdPdf product{"product", {&exp_x, &exp_y}};

    product.setData(&data);
    FitManager fitter(&product);
    fitter.fit();

    return fitter;
}
